### PR TITLE
Add multicast subcommand

### PR DESCRIPTION
**Pull request by yushoyamaguchi**
## Change
This change adds cilium multicast sub command to manage multicast groups and subscribers throughout the cluster

Following commands are added
```
# show list of multicast group
- cilium multicast list group

# show list of subscribers
- cilium multicast list subscriber --all
- cilium multicast list subscriber --group-ip <group-ip>

# make specified multicast group and add CiliumInternalIP of all cilium nodes to the group in each node
- cilium multicast add --group-ip <group-ip>

# Remove CiliumInternalIP of all cilium nodes from specified multicast group and delete the group in each node
- cilium multicast del --group-ip <group-ip>
```
This is follow up of https://github.com/cilium/cilium-cli/issues/2574

## Display example
cli-result

## Additional
As for testing, we will proceed with the implementation of the connectivity test for multicast as described in this issue.(https://github.com/cilium/cilium-cli/issues/2615)

## Reference
Current way of setup multicast : https://github.com/cilium/cilium/blob/main/Documentation/network/multicast.rst

### CODE REVIEW

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:217]: Adding pod name in this message would be helpful.

File: cli/multicast.go
Comment (harsimran-pabla) [inline cli/multicast.go:95]: I'd refrain from mentioning CiliumInternalIPs in the description. It is an internal detail; a user should not care about it.

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:235]: Consider creating a helper method for this; the same pattern is used in the add and del methods.

File: multicast/multicast.go
Comment (harsimran-pabla) [reply multicast/multicast.go:212]: If you call this CLI multiple times - you'll receive an 'already exists' type of error. If that is the case, check the error type. The 'already exists' error can be ignored ( idempotent call ), and for any other error type, we fail and return.

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:246]: similar comment: print should be removed.

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:255]: It might be better to check error type.

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:266]: Similar comment: need more robust error handling.

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:271]:
```suggestion
		fmt.Fprintf(m.params.Writer, "Deleted %d subscribers from multicast group %s\n\n", cnt, m.params.MulticastGroupIP)
```

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:273]: Maybe all the subscriber cleanup is unnecessary; just deleting the group should be enough.

**Review by harsimran-pabla — CHANGES_REQUESTED** [review]
Thank you for the changes; overall, it looks good. A couple of areas that need work are concurrency and error handling.

**Comment by yushoyamaguchi** [comment]
@harsimran-pabla 
Thank you so much for reviewing.

I'm sorry, I have a question.
github_err_msg
These warning messages come from GitHub actions.
In other subcommands, variable name `cmd` is used.  Can I ignore these warning messages, or should I change variable name to `_` ?

File: cli/multicast.go
Comment (yushoyamaguchi) [reply cli/multicast.go]: I've fixed

**Review by asauber — COMMENTED** [review]
I will save a detailed review for after the current round of changes.

**Comment by harsimran-pabla** [comment]
Thanks @yushoyamaguchi  for addressing these comments.

> In other subcommands, variable name cmd is used. Can I ignore these warning messages, or should I change variable name to _ ?

It would be best if you fixed these comments ( these are static syntax checkers and have to make linters happy ). 

>Second, setting resolved tag for the place I fixed is my role or your role? (Currently, I only put 👍)
I'm sorry for ignorance about OSS development.

If you have addressed the comment; feel free to resolve it. 

> In addition, I replied to some your fix request comments which I cannot find the way to fix.

Please add a comment whichever comment is not clear, I will try my best to explain it further.

**Comment by harsimran-pabla** [comment]
General commit comment. 

- Keep each commit about logical change instead of iterations on this PR. 
- Add more details about the change in the commit description.

File: multicast/multicast.go
Comment (asauber) [inline multicast/multicast.go:164]: nit: If you are going to leave this `//ToDo: ` here, which is okay, it should be formatted as `// TODO `. Same for other occurrences of `ToDo`, also inline comments should start like `// <comment>` rather than `//<comment>`.

File: multicast/multicast.go
Comment (asauber) [inline multicast/multicast.go:133]: This should be named `ListSubscribers`

File: multicast/multicast.go
Comment (asauber) [inline multicast/multicast.go:92]: This should be named `ListGroups`

File: multicast/multicast.go
Comment (asauber) [inline multicast/multicast.go:232]: I'd recommend splitting this into a new function, if not purely for the sake of readability and avoiding the pitfalls of `wg1, wg2..`

**Review by asauber — CHANGES_REQUESTED** [review]

**Comment by yushoyamaguchi** [comment]
@asauber 
Thank you for your reviewing.
I'll fix all improvement point, then tell you.

**Comment by yushoyamaguchi** [comment]
@harsimran-pabla @asauber 

Thank you so much for supporting.
I've finished fixing all points.

I'm sorry for making git log complicated, but all commits are grouped by meaning.
Only minor changes are put in [this commit](https://github.com/cilium/cilium-cli/pull/2620/commits/66f6108173f2db10e89487793f53966e83c117f7) by force push.

**Review by asauber — CHANGES_REQUESTED** [review]
This is shaping up nicely. Thank you for the updates.

Would you mind squashing all of the commits which modify your own code introduced in this PR?

The purpose of individual commits in a ready-to-merge PR should be to introduce new functionality, each with a detailed description. (For an example of a description, we have this recent commit https://github.com/cilium/cilium-cli/pull/2569/commits/0dea3f7de7c67775c29f452fdcdce897f89a9486)

You can feel free to break up the PR into "additive" commits if you wish, for example, one commit for each newly added command.

**Comment by yushoyamaguchi** [comment]
@asauber 
I combined all the changes into one commit by `git squash` and explained the changes in detail in the commit message.
https://github.com/cilium/cilium-cli/pull/2620/commits/a923c64cb8f53b10b81d88a6abe1a7fc146230e9

cc @harsimran-pabla

**Review by asauber — COMMENTED** [review]
Thanks, looks good to me.

In order to fix the failing tests (required for merge), can you try rebasing this branch on main?

**Comment by yushoyamaguchi** [comment]
@asauber 
Thank you for your reviewing.
I 've fixed some errors from golang compiler and rebased this branch on main.

**Review by asauber — COMMENTED** [review]
Looks like this needs one more round of lint fixes

**Comment by yushoyamaguchi** [comment]
@asauber 
All checks have passed.
Thank you.

File: multicast/multicast.go
Comment (asauber) [inline multicast/multicast.go:222]: For tabular output, we prefer to use the `tabwriter` package for alignment.

See https://github.com/cilium/cilium-cli/blob/main/bgp/routes.go#L257 for an example.

File: multicast/multicast.go
Comment (asauber) [inline multicast/multicast.go:329]: nit: another pass for `// ` (space after comment marker), would be appreciated

**Review by asauber — CHANGES_REQUESTED** [review]

**Comment by yushoyamaguchi** [comment]
@asauber 
Thank you for reviewing.
I've fixed all points.

In [here](https://github.com/cilium/cilium-cli/pull/2620/files#diff-abe2c01976e393eb320f6c1520249d16541f969e33771b03acf3516d6fa3008cR186), there is a comment that it depends on the cilium-dbg code.


Since it is impossible in principle to eliminate the dependence on cilium-dbg in this program, I thought it would be a good idea to put a comment in both cilium-dbg and cilium-dbg where they affect each other.

**Review by asauber — APPROVED** [review]
very nice, thanks for the updates

File: cli/multicast.go
Comment (harsimran-pabla) [inline cli/multicast.go:117]: nit: del -> delete. `del` can be added as an alias.

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:107]: I have a common concern regarding all these commands. Output is in custom tabular format, which is not ideal as some users may want to consume it in other formats, like JSON, to automate some workflows. 

Could you consider separating data collection and data output in these methods? Fetch data from all the nodes, then output data based on the user's specified format.

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:114]: Consider passing `-o json` to cilium-dbg command, it would make data parsing much easier ( use json unmarshaller ).

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:229]: Similar comment regarding usage of json data from cilium-dbg.

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:390]: Is this data collection required? I do not see it being used anywhere.

**Review by harsimran-pabla — CHANGES_REQUESTED** [review]
Thanks, @yushoyamaguchi; change is coming along nicely. I have a couple of suggestions.

- Consider a use case where someone might want JSON format to display output. 
- When fetching data from cilium pods, use `-o json`; it makes it much easier to parse this information.

**Comment by yushoyamaguchi** [comment]
@harsimran-pabla 

Thank you for reviewing.

> Consider a use case where someone might want JSON format to display output.
> When fetching data from cilium pods, use -o json; it makes it much easier to parse this information.

I have addressed both of these requirements.
The output format is not changed.

Because the amount of code changes is significant, I split them into separate commits for better visibility.
If you prefer to have them in one commit when merging, please say so.

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:62]:
```suggestion
type GroupSubscriberData struct {
```

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:67]:
```suggestion
type NodeSubscriberData struct {
```

File: multicast/multicast.go
Comment (harsimran-pabla) [inline multicast/multicast.go:72]:
```suggestion
type NodeGroupData struct {
```

**Review by harsimran-pabla — CHANGES_REQUESTED** [review]
Thanks @yushoyamaguchi, for bearing with me on this PR. There are a couple of things I would like to highlight: minor naming changes. And channel creations can be cleaned up. I'd suggest consolidating the channels, as shown in the example in one of the comments.

**Comment by yushoyamaguchi** [comment]
@harsimran-pabla 

Thank you for kindly pointing out the places that needed improvement.
I'm happy because I could learn a lot.

I've fixed all the points you mentioned.

**Review by harsimran-pabla — APPROVED** [review]

**Comment by yushoyamaguchi** [comment]
@harsimran-pabla 
Thank you so much for the detailed review
I've squashed.

**Comment by yushoyamaguchi** [comment]
@harsimran-pabla @asauber 
Thank you for the approval.
Should I do something else for merging?
As I said in [this PR](https://github.com/cilium/cilium-cli/pull/2724#issue-2430113723), I want to add test scenario using this multicast subcommand after merging...

-----
@fujitatomoya -san 's approval is required?
He said maybe not required but I can ask him if needed, because he is a my colleague.

**Comment by harsimran-pabla** [comment]
Thanks, @yushoyamaguchi. I think this PR is ready to merge; I believe all concerns are addressed. We can move forward with e2e testing independently.

**Comment by harsimran-pabla** [comment]
I'd suggest reaching out to @fujitatomoya to get his approval as well, since he reviewed this PR.

**Comment by yushoyamaguchi** [comment]
Thank you.

**Review by fujitatomoya — APPROVED** [review]
@harsimran-pabla @asauber thanks for the reviews, im good to go.

**Review by michi-covalent — APPROVED** [review]

**Comment by yushoyamaguchi** [comment]
@harsimran-pabla @asauber @michi-covalent @fujitatomoya 
Thank you so much!
Should I write some Document?

**Pull request merged**
